Death to pal::RTree

nyalldawson opened the pull request:
This PR drops to ancient, unmaintained, custom pal::RTree spatial index and replaces it with a backend utilising libspatialindex. This brings it into line with the rest of QGIS, which heavily uses libspatialindex, and removes a couple of thousand lines of totally maintainable code. 

(Possibly there's a speed boost because of this too, but the original rtree code never took a significant amount of time in the context of label problem solving anyway...)

nyalldawson committed:
Start moving some spatial index related functions to QgsSpatialIndexUtils

nyalldawson committed:
Start on QgsGenericSpatialIndex to replace pal rtree with libspatialindex based backend

nyalldawson committed:
More pal::rtree -> QgsGenericSpatialIndex ports

nyalldawson committed:
More pal::rtree swaps

nyalldawson committed:
Remove last use of pal::RTree

nyalldawson committed:
Death to pal::RTree

nyalldawson committed:
Drop some more unused pal code

NathanW2 commented:
So long pal::RTree

On Mon, Dec 9, 2019 at 11:53 AM Nyall Dawson <notifications@github.com>
wrote:

> This PR drops to ancient, unmaintained, cust pal::RTree spatial index and
> replaces it with a backend utilising libspatialindex. This brings it into
> line with the rest of QGIS, which heavily uses libspatialindex, and removes
> a couple of thousand lines of totally maintainable code.
>
> (Possibly there's a speed boost because of this too, but the original
> rtree code never took a significant amount of time in the context of label
> problem solving anyway...)
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/qgis/QGIS/pull/33288
> Commit Summary
>
>    - Start moving some spatial index related functions to
>    QgsSpatialIndexUtils
>    - Start on QgsGenericSpatialIndex to replace pal rtree with
>    libspatialindex based backend
>    - More pal::rtree -> QgsGenericSpatialIndex ports
>    - More pal::rtree swaps
>    - Remove last use of pal::RTree
>    - Death to pal::RTree
>    - Drop some more unused pal code
>
> File Changes
>
>    - *M* python/core/auto_generated/qgslabelsearchtree.sip.in
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-0> (7)
>    - *M* src/core/CMakeLists.txt
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-1> (4)
>    - *M* src/core/pal/costcalculator.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-2> (24)
>    - *M* src/core/pal/costcalculator.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-3> (11)
>    - *M* src/core/pal/feature.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-4> (1)
>    - *M* src/core/pal/labelposition.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-5> (96)
>    - *M* src/core/pal/labelposition.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-6> (24)
>    - *M* src/core/pal/layer.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-7> (39)
>    - *M* src/core/pal/layer.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-8> (10)
>    - *M* src/core/pal/pal.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-9> (240)
>    - *D* src/core/pal/palgeometry.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-10> (40)
>    - *M* src/core/pal/pointset.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-11> (13)
>    - *M* src/core/pal/problem.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-12> (254)
>    - *M* src/core/pal/problem.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-13> (6)
>    - *D* src/core/pal/rtree.hpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-14> (1598)
>    - *M* src/core/pal/util.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-15> (47)
>    - *M* src/core/pal/util.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-16> (8)
>    - *A* src/core/qgsgenericspatialindex.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-17> (153)
>    - *A* src/core/qgsgenericspatialindex.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-18> (58)
>    - *M* src/core/qgslabelsearchtree.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-19> (52)
>    - *M* src/core/qgslabelsearchtree.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-20> (14)
>    - *M* src/core/qgsspatialindex.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-21> (16)
>    - *M* src/core/qgsspatialindex.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-22> (2)
>    - *A* src/core/qgsspatialindexutils.cpp
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-23> (25)
>    - *A* src/core/qgsspatialindexutils.h
>    <https://github.com/qgis/QGIS/pull/33288/files#diff-24> (60)
>
> Patch Links:
>
>    - https://github.com/qgis/QGIS/pull/33288.patch
>    - https://github.com/qgis/QGIS/pull/33288.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/33288?email_source=notifications&email_token=AAC5FXBBKICEDP4IHZHQLS3QXWQIHA5CNFSM4JYB6M62YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4H66BNIA>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXGSBRIQAEOLVIL7HADQXWQIHANCNFSM4JYB6M6Q>
> .
>

nyalldawson committed:
Simplify code

nyalldawson committed:
Simplify code

nyalldawson committed:
Simplify code and remove dead code

nyalldawson committed:
Remove dead code

nyalldawson committed:
Dox, nicer naming

nyalldawson committed:
Avoid ugly const casts

nirvn commented:
I can't pretend do be in a position to do a proper review of all of this code, so I wont. That said, I've compiled the branch, and tested a couple of my label-heavy projects, and happy to report I can't spot regressions. Nice to see over 2,000 line of codes going away too.

nyalldawson commented:
@nirvn thanks for the testing -- it's certainly appreciated/valuable!

nyalldawson committed:
Dox++

nyalldawson committed:
Stop fighting with templates and take the shortcut approach

nyalldawson committed:
Unit tests for QgsGenericSpatialIndex

nyalldawson committed:
Spelling

nyalldawson committed:
Don't use a recursive mutex

Earlier versions of libspatialindex on linux had their own internal
(non-recursive) mutex, so to behavior consistent across different environments
we need to ensure we can survive without a non-recursive external mutex

nyalldawson committed:
Avoid recursive index calls

nyalldawson committed:
Update some label reference images

nyalldawson committed:
More test image updates

nyalldawson committed:
Test updates

nyalldawson committed:
More test updates

nyalldawson committed:
Use a report for selective mask tests, try to make tests stable

nyalldawson committed:
Add test masks

nyalldawson committed:
Don't try to build genericspatialindex test on windows

Can't find the magic formula to make msvc happy here!

nyalldawson committed:
Avoid some unnecessary label index construction

The pal library was originally designed so that features would be
registered once and used across multiple map redraws. We don't utilise
that, and we are gauranteed that all registered features and obstacles
already fall within the desired map boundary.

nyalldawson committed:
Remove unused variables

nyalldawson committed:
Indentation

nyalldawson committed:
Test updates again

nyalldawson committed:
More test updates
